src/main/java/emissary/core/ResourceWatcher.java

```diff
@@ -148,15 +151,23 @@ public void run() {
 
     public void logStats(final Logger loggerArg) {
         for (final Map.Entry<String, Timer> e : this.metrics.getTimers().entrySet()) {
-            loggerArg.info(this.metricsFormatter.formatTimer(e.getKey(), e.getValue()));
+            // We only want to log stats for places that have had events
+            if (e.getValue().getCount() > 0) {
+                loggerArg.info(this.metricsFormatter.formatTimer(e.getKey(), e.getValue()));
+            }
         }
     }
 
     public void resetStats() {
-        // no reset but remove:
-        // https://github.com/codahale/metrics/issues/399
-        for (final Map.Entry<String, Timer> e : this.metrics.getTimers().entrySet()) {
-            this.metrics.remove(e.getKey());
+        // We use reflection to reset the histograms that track finished events, but leaves the namespace for active timers
+        for (Timer timer : this.metrics.getTimers().values()) {
+            try {
+                Field histogramField = Timer.class.getDeclaredField("histogram");
+                histogramField.setAccessible(true);
+                histogramField.set(timer, new Histogram(new ExponentiallyDecayingReservoir()));
```

Comment:
Is reflection the only way to do this?  Is there anything in the new version that fixes what we are trying to solve? Is there a different library that we could use? I'm just wondering if we are using the library is a way that it wasn't intended, so is there another option that would be better long term? Also, I'm afraid of a change like this affecting us being able to upgrade to a new version in the future.

Reply:
@dev-mlb 
I have done a second due-dillegence check. The newest version of this library has the same limitations and there are no ways other than reflection to accomplish the desired effect.

There is a rolling-metric library developed by a third party (https://github.com/vladimir-bukhtoyarov/rolling-metrics) that would allow resetting histograms when we perform the getSnapshot() in the MetricsFormatter. However, this solution adds two additional dependencies to our project and it is unclear what future maintenance for these libraries will look like.

Histograms as part of the timers has been a core part of the library infrastructure since inception. This structure would be seemingly unlikely to change and shouldn't prevent an upgrade. Any upgrade issues could be easily resolved by updating the reflection usage or implementing any new functions that allow the desired behavior.

I added a new line to the ResourceWatcherTest to ensure that the reset occurs as expected.

I recommend proceeding with the current changes. Any alternative solutions will be time-intensive to implement and in my opinion incur just as much if not more risk and challenge for future upgrades.